src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/RestoreCommandHandler.cs

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.ServiceHub.Framework;
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Shell.ServiceBroker;
+using Microsoft.VisualStudio.Threading;
+using NuGet.VisualStudio.Internal.Contracts;
+using IAsyncServiceProvider = Microsoft.VisualStudio.Shell.IAsyncServiceProvider;
+using Task = System.Threading.Tasks.Task;
+
+namespace NuGet.VisualStudio.OnlineEnvironment.Client
+{
+    internal class RestoreCommandHandler
+    {
+        private readonly JoinableTaskFactory _joinableTaskFactory;
+        private readonly IAsyncServiceProvider _asyncServiceProvider;
+        private CancellationTokenSource _cancelBuildToken;
+
+        public RestoreCommandHandler(JoinableTaskFactory joinableTaskFactory, IAsyncServiceProvider asyncServiceProvider)
+        {
+            _joinableTaskFactory = joinableTaskFactory ?? throw new ArgumentNullException(nameof(joinableTaskFactory));
+            _asyncServiceProvider = asyncServiceProvider ?? throw new ArgumentNullException(nameof(asyncServiceProvider));
+        }
+
+        public void RunSolutionRestore()
+        {
+            _joinableTaskFactory.RunAsync(async () =>
+            {
+                var cancellationTokenSource = new CancellationTokenSource();
```

Comment:
I'm not quite following the use of 2 vars for this obj. Can you just use the field?

Reply:
It's a cheap synchronization approach. 
When scripting this could somehow get called twice, and you'd leak a cancellationtokensource. 

I'm sure there's a 100 ways to do this.